File: src/main/java/main/utils/pagination/PaginationHandler.java
package main.utils.pagination;

import com.github.topisenpai.lavasrc.mirror.MirroringAudioTrack;
import lombok.SneakyThrows;
import main.audiohandlers.QueueHandler;
import main.audiohandlers.RobertifyAudioManager;
import main.commands.prefixcommands.audio.QueueCommand;
import main.constants.InteractionLimits;
import main.utils.RobertifyEmbedUtils;
import main.utils.apis.robertify.imagebuilders.AbstractImageBuilder;
import main.utils.apis.robertify.imagebuilders.ImageBuilderException;
import main.utils.component.interactions.selectionmenu.StringSelectMenuOption;
import main.utils.component.interactions.selectionmenu.StringSelectionMenuBuilder;
import main.utils.pagination.pages.DefaultMessagePage;
import main.utils.pagination.pages.MenuPage;
import main.utils.pagination.pages.MessagePage;
import main.utils.pagination.pages.queue.QueueItem;
import main.utils.pagination.pages.queue.QueuePage;
import net.dv8tion.jda.api.EmbedBuilder;
import net.dv8tion.jda.api.entities.Guild;
import net.dv8tion.jda.api.entities.Message;
import net.dv8tion.jda.api.entities.MessageEmbed;
import net.dv8tion.jda.api.entities.User;
import net.dv8tion.jda.api.entities.channel.middleman.GuildMessageChannel;
import net.dv8tion.jda.api.events.interaction.command.SlashCommandInteractionEvent;
import net.dv8tion.jda.api.interactions.components.ActionRow;
import net.dv8tion.jda.api.interactions.components.selections.SelectMenu;
import net.dv8tion.jda.api.requests.restaction.WebhookMessageCreateAction;
import net.dv8tion.jda.api.requests.restaction.interactions.ReplyCallbackAction;
import net.dv8tion.jda.api.utils.FileUpload;
import net.dv8tion.jda.internal.utils.tuple.Pair;

import java.net.ConnectException;
import java.net.SocketTimeoutException;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.concurrent.atomic.AtomicReference;
import java.util.function.Supplier;

public abstract class PaginationHandler {
    private static final HashMap<Long, List<? extends MessagePage>> messages = new HashMap<>();
    private static Supplier<EmbedBuilder> embedStyle = EmbedBuilder::new;

    public static Message paginateMessage(GuildMessageChannel channel, User user, List<DefaultMessagePage> messagePages) {
        AtomicReference<Message> ret = new AtomicReference<>();

        channel.sendMessageEmbeds(messagePages.get(0).getEmbed()).queue(msg -> {
            if (messagePages.size() > 1) {
                msg.editMessageComponents(
                        Paginator.getButtons(user, false, false, true, true)
                ).queue();

                messages.put(msg.getIdLong(), messagePages);
                ret.set(msg);
            }
        });

        return ret.get();
    }

    public static Message paginateMessage(SlashCommandInteractionEvent event, List<DefaultMessagePage> messagePages) {
        AtomicReference<Message> ret = new AtomicReference<>();

        WebhookMessageCreateAction<Message> messageAction = event.getHook().sendMessageEmbeds(messagePages.get(0).getEmbed()).setEphemeral(RobertifyEmbedUtils.getEphemeralState(event.getChannel().asGuildMessageChannel()));

        if (messagePages.size() > 1) {
            messageAction = messageAction.addComponents(
                    Paginator.getButtons(event.getUser(), false, false, true, true)
            );
        }

        messageAction.queue(msg -> {
            if (messagePages.size() > 1) {
                messages.put(msg.getIdLong(), messagePages);
                ret.set(msg);
            }
        });

        return ret.get();
    }

    public static Message paginateQueueMessage(SlashCommandInteractionEvent event, List<QueuePage> queuePages) {
        AtomicReference<Message> ret = new AtomicReference<>();
        assert event.getGuild() != null;

        try {
            final var image = queuePages.get(0).getImage();
            WebhookMessageCreateAction<Message> messageAction = event.getHook()
                    .sendFiles(FileUpload.fromData(image, AbstractImageBuilder.getRandomFileName()))
                    .setEphemeral(RobertifyEmbedUtils.getEphemeralState(event.getChannel().asGuildMessageChannel()));

            if (queuePages.size() > 1) {
                messageAction = messageAction.addComponents(
                        Paginator.getQueueButtons(event.getUser(), false, false, true, true)
                );
            }

            messageAction.queue(msg -> {
                if (queuePages.size() > 1) {
                    messages.put(msg.getIdLong(), queuePages);
                    ret.set(msg);
                }
            });
        } catch (SocketTimeoutException | ConnectException e) {
            final var pages = new ArrayList<DefaultMessagePage>();
            final var musicManager = RobertifyAudioManager.getInstance().getMusicManager(event.getGuild());
            final var queueHandler = musicManager.getScheduler().getQueueHandler();
            final var content = new QueueCommand().getPastTrackContent(event.getGuild(), queueHandler);

            messageLogic(event.getGuild(), pages, content, 10);
            paginateMessage(event, pages);
        }


        return ret.get();
    }

    public static Message paginateMessage(GuildMessageChannel channel, User user, List<String> content, int maxPerPage) {
        List<DefaultMessagePage> messagePages = new ArrayList<>();

        messageLogic(channel.getGuild(), messagePages, content, maxPerPage);

        return paginateMessage(channel, user, messagePages);
    }

    public static Message paginateMessage(List<String> content, int maxPerPage, SlashCommandInteractionEvent event) {
        List<DefaultMessagePage> messagePages = new ArrayList<>();

        event.deferReply().queue();

        messageLogic(event.getGuild(), messagePages, content, maxPerPage);

        return paginateMessage(event, messagePages);
    }

    public static Message paginateQueue(int maxPerPage, SlashCommandInteractionEvent event) {
        final var musicManager = RobertifyAudioManager.getInstance().getMusicManager(event.getGuild());
        final var queueHandler = musicManager.getScheduler().getQueueHandler();
        final var queuePages = new ArrayList<QueuePage>();

        event.deferReply().queue();

        messageLogic(event.getGuild(), queuePages, queueHandler, maxPerPage);

        try {
            return paginateQueueMessage(event, queuePages);
        } catch (ImageBuilderException e) {
            final var defaultMessagePages = queuePages.stream()
                    .map(page -> new DefaultMessagePage(page.getEmbed()))
                    .toList();
           return paginateMessage(event, defaultMessagePages);
        }
    }



    private static void messageLogic(Guild guild, List<DefaultMessagePage> messagePages, List<String> content, int maxPerPage) {
        if (content.size() <= maxPerPage) {
            messagePages.add(new DefaultMessagePage(guild, content));
        } else {
            int pagesRequired = (int) Math.ceil((double) content.size() / maxPerPage);

            int lastIndex = 0;
            for (int i = 0; i < pagesRequired; i++) {
                EmbedBuilder eb = RobertifyEmbedUtils.embedMessage(guild, "\t");
                for (int j = 0; j < maxPerPage; j++) {
                    if (lastIndex == content.size()) break;

                    eb.appendDescription(content.get(lastIndex) + "\n");
                    lastIndex++;
                }
                messagePages.add(new DefaultMessagePage(eb.build()));
            }
        }
    }

    private static void messageLogic(Guild guild, List<QueuePage> messagePages, QueueHandler queueHandler, int maxPerPage) {
        if (queueHandler.size() <= maxPerPage) {
            final List<QueueItem> items = new ArrayList<>();
            for (int i = 0; i < queueHandler.size(); i++) {
                final var track = queueHandler.contents().get(i);
                final var trackInfo = track.getInfo();
                items.add(new QueueItem(i + 1, trackInfo.title, trackInfo.author, trackInfo.length, track instanceof MirroringAudioTrack mt ? mt.getArtworkURL() : null));
            }
            messagePages.add(new QueuePage(guild, 1, items));
        } else {
            final var trackList = queueHandler.contents();
            int pagesRequired = (int)Math.ceil((double)queueHandler.size() / maxPerPage);
            int lastIndex = 0;

            for (int i = 0; i < pagesRequired; i++) {
                final var page = new QueuePage(guild, i + 1);
                for (int j = 0; j < maxPerPage; j++) {
                    if (lastIndex == queueHandler.size())
                        break;
                    final var track = trackList.get(lastIndex);
                    final var trackInfo = track.getInfo();

                    page.addItem(lastIndex + 1, trackInfo.title, trackInfo.author, trackInfo.length, track instanceof MirroringAudioTrack mt ? mt.getArtworkURL() : null);
                    lastIndex++;
                }
                messagePages.add(page);
            }
        }
    }

    public static List<? extends MessagePage> getMessagePages(long msg) {
        return messages.get(msg);
    }

    public static List<QueuePage> getQueuePages(long msg) {
        return (List<QueuePage>) getMessagePages(msg);
    }

    @SneakyThrows
    public static void paginateMenu(User user, Message msg, List<StringSelectMenuOption> options) {
        List<MenuPage> menuPages = menuLogic(msg.getId(), options);

        final var firstPage = menuPages.get(0);

        SelectMenu menu = StringSelectionMenuBuilder.of(
                "menupage:" + user.getId(),
                "Select an option",
                Pair.of(1, 1),
                firstPage.getOptions().subList(0, Math.min(options.size(), InteractionLimits.SELECTION_MENU))
        ).build();

        msg.editMessageComponents(ActionRow.of(menu))
                .queue(success -> messages.put(msg.getIdLong(), menuPages));
    }

    @SneakyThrows
    public static void paginateMenu(User user, GuildMessageChannel channel, ReplyCallbackAction msg, List<StringSelectMenuOption> options) {
        List<MenuPage> menuPages = menuLogic("null", options);

        final var firstPage = menuPages.get(0);

        SelectMenu menu = StringSelectionMenuBuilder.of(
                "menupage:" + user.getId(),
                "Select an option",
                Pair.of(1, 1),
                firstPage.getOptions().subList(0, Math.min(options.size(), InteractionLimits.SELECTION_MENU))
        ).build();

        msg.addActionRow(menu)
                .setEphemeral(RobertifyEmbedUtils.getEphemeralState(channel))
                .queue(success -> success.retrieveOriginal().queue(og -> messages.put(og.getIdLong(), menuPages)));
    }

    public static void paginateMenu(GuildMessageChannel channel, User user, List<StringSelectMenuOption> options, int startingPage) {
        paginateMenu(channel, user, options, startingPage, false);
    }

    public static void paginateMenu(GuildMessageChannel channel, User user, List<StringSelectMenuOption> options, int startingPage, boolean numberEachEntry) {
        Message msg = menuLogic(channel, options, startingPage, numberEachEntry);
        paginateMenu(user, msg, options);
    }

    public static void paginateMenu(SlashCommandInteractionEvent event, List<StringSelectMenuOption> options, int startingPage, boolean numberEachEntry) {
        final var msg = menuLogic(event, options, startingPage, numberEachEntry);
        paginateMenu(event.getUser(), event.getChannel().asGuildMessageChannel(), msg, options);
    }

    private static Message menuLogic(GuildMessageChannel channel, int startingPage, List<StringSelectMenuOption> options) {
        return menuLogic(channel, options, startingPage, false);
    }

    private static ReplyCallbackAction menuLogic(SlashCommandInteractionEvent event, int startingPage, List<StringSelectMenuOption> options) {
        return menuLogic(event, options, startingPage, false);
    }

    private static Message menuLogic(GuildMessageChannel channel, List<StringSelectMenuOption> options, int startingPage, boolean numberEachEntry) {
        return channel.sendMessageEmbeds(getPaginatedEmbed(channel.getGuild(), options, 25, startingPage, numberEachEntry)).complete();
    }

    private static ReplyCallbackAction menuLogic(SlashCommandInteractionEvent event, List<StringSelectMenuOption> options, int startingPage, boolean numberEachEntry) {
        return event.replyEmbeds(getPaginatedEmbed(event.getGuild(), options, 25, startingPage, numberEachEntry));
    }

    private static List<MenuPage> menuLogic(String msgID, List<StringSelectMenuOption> options) {
        final List<MenuPage> menuPages = new ArrayList<>();

        if (options.size() <= InteractionLimits.SELECTION_MENU) {
            MenuPage menuPage = new MenuPage();
            for (final var option : options)
                menuPage.addOption(option);
            menuPages.add(menuPage);
        } else {

            final int pagesRequired = (int) Math.ceil((double) options.size() / InteractionLimits.SELECTION_MENU);
            final int pageControllers = 1 + (int) Math.ceil(((pagesRequired - 1) * (4)) / 2.0);
            final int actualPagesRequired = (int) Math.ceil((double) (options.size() + pageControllers) / InteractionLimits.SELECTION_MENU);

            int lastIndex = 0;
            for (int i = 0; i < actualPagesRequired; i++) {
                final MenuPage tempPage = new MenuPage();
                for (int j = 0; j < InteractionLimits.SELECTION_MENU; j++) {
                    if (lastIndex == options.size()) break;

                    if (j == 0 && i != 0) {
                        tempPage.addOption(StringSelectMenuOption.of("Previous Page", "menuPage:previousPage:" + msgID));
                        continue;
                    }

                    if (j == InteractionLimits.SELECTION_MENU - 1) {
                        tempPage.addOption(StringSelectMenuOption.of("Next Page", "menuPage:nextPage:" + msgID));
                        continue;
                    }

                    tempPage.addOption(options.get(lastIndex));
                    lastIndex++;
                }
                menuPages.add(tempPage);
            }
        }

        return menuPages;
    }

    public static List<MenuPage> getMenuPages(long msg) {
        return (List<MenuPage>) messages.get(msg);
    }

    @SneakyThrows
    public static SelectMenu getSelectionMenu(User user, List<StringSelectMenuOption> options) {
        return StringSelectionMenuBuilder.of(
                "menuPage:" + user.getIdLong(),
                "Select an option",
                Pair.of(1, 1),
                options.subList(0, Math.min(options.size(), InteractionLimits.SELECTION_MENU))
        ).build();
    }

    public static MessageEmbed getPaginatedEmbed(Guild guild, List<?> content, int maxPerPage, int startingPage) {
        return getPaginatedEmbed(guild, content, maxPerPage, startingPage, false);
    }

    public static void setEmbedStyle(Supplier<EmbedBuilder> supplier) {
        embedStyle = supplier;
    }

    public static MessageEmbed getPaginatedEmbed(Guild guild, List<?> content, int maxPerPage, int startingPage, boolean numberEachEntry) {
        EmbedBuilder eb = embedStyle.get().appendDescription("\t");

        int index = (startingPage * (maxPerPage == 0 ? maxPerPage : maxPerPage - 1)) + 1;
        for (int j = 0; j < content.size(); j++) {
            if (j == maxPerPage) break;

            eb.appendDescription((numberEachEntry ? "**" + (index++) + ".** - " : "") + content.get(j) + "\n");
        }

        return eb.build();
    }

}

File: src/main/kotlin/main/utils/pagination/PaginationHandler.kt
package main.utils.pagination

import dev.minn.jda.ktx.interactions.components.StringSelectMenu
import dev.minn.jda.ktx.util.SLF4J
import kotlinx.coroutines.Deferred
import kotlinx.coroutines.async
import kotlinx.coroutines.coroutineScope
import main.audiohandlers.QueueHandler
import main.audiohandlers.RobertifyAudioManager
import main.audiohandlers.utils.author
import main.audiohandlers.utils.length
import main.audiohandlers.utils.title
import main.constants.InteractionLimits
import main.utils.GeneralUtils.coerceAtMost
import main.utils.RobertifyEmbedUtils
import main.utils.RobertifyEmbedUtils.Companion.sendEmbed
import main.utils.api.robertify.imagebuilders.AbstractImageBuilder
import main.utils.api.robertify.imagebuilders.ImageBuilderException
import main.utils.component.interactions.selectionmenu.StringSelectMenuOption
import main.utils.component.interactions.selectionmenu.StringSelectionMenuBuilder
import main.utils.database.mongodb.databases.playlists.PlaylistDB
import main.utils.database.mongodb.databases.playlists.PlaylistModel
import main.utils.database.mongodb.databases.playlists.PlaylistTrack
import main.utils.pagination.pages.AbstractImagePage
import main.utils.pagination.pages.DefaultMessagePage
import main.utils.pagination.pages.MenuPage
import main.utils.pagination.pages.MessagePage
import main.utils.pagination.pages.playlists.PlaylistPage
import main.utils.pagination.pages.playlists.PlaylistsPage
import main.utils.pagination.pages.queue.QueueItem
import main.utils.pagination.pages.queue.QueuePage
import net.dv8tion.jda.api.EmbedBuilder
import net.dv8tion.jda.api.entities.Guild
import net.dv8tion.jda.api.entities.Message
import net.dv8tion.jda.api.entities.MessageEmbed
import net.dv8tion.jda.api.entities.User
import net.dv8tion.jda.api.entities.channel.middleman.GuildMessageChannel
import net.dv8tion.jda.api.events.interaction.command.SlashCommandInteractionEvent
import net.dv8tion.jda.api.interactions.components.ActionRow
import net.dv8tion.jda.api.interactions.components.selections.SelectMenu
import net.dv8tion.jda.api.requests.restaction.interactions.ReplyCallbackAction
import net.dv8tion.jda.api.utils.FileUpload
import java.net.ConnectException
import java.net.SocketTimeoutException
import java.util.*
import kotlin.math.ceil

object PaginationHandler {
    private val logger by SLF4J
    private val messages = Collections.synchronizedMap(mutableMapOf<Long, List<MessagePage>>())
    private val paginator = PaginationButtonGenerator.DEFAULT
    var embedStyle: suspend () -> EmbedBuilder = { EmbedBuilder() }

    suspend fun paginateMessage(channel: GuildMessageChannel, user: User, messagePages: List<MessagePage>) =
        coroutineScope {
            val paginatedMessage = async {
                channel.sendMessageEmbeds(messagePages[0].getEmbed()!!)
                    .submit()
                    .thenApply { message ->
                        if (messagePages.size > 1) {
                            message.editMessageComponents(
                                paginator.getButtons(
                                    user = user,
                                    frontEnabled = false,
                                    previousEnabled = false
                                )
                            ).queue()

                            messages[message.idLong] = messagePages
                            return@thenApply message
                        } else return@thenApply null
                    }.join()
            }
            return@coroutineScope paginatedMessage.await()
        }

    suspend fun paginateMessage(
        event: SlashCommandInteractionEvent,
        messagePages: List<MessagePage>,
        isQueue: Boolean = false
    ): Message? =
        coroutineScope {
            val paginatedMessage = async {
                var messageAction = event.hook.sendEmbed(event.guild) { messagePages[0].getEmbed()!! }

                if (messagePages.size > 1)
                    messageAction = messageAction.addComponents(
                        paginator.getButtons(
                            user = event.user,
                            frontEnabled = false,
                            previousEnabled = false,
                            isQueue = isQueue
                        )
                    )

                return@async messageAction.submit()
                    .thenApply { msg ->
                        if (messagePages.size > 1) {
                            messages[msg.idLong] = messagePages
                            return@thenApply msg
                        } else return@thenApply null
                    }.join()
            }
            return@coroutineScope paginatedMessage.await()
        }

    suspend fun paginateMessage(
        event: SlashCommandInteractionEvent,
        content: List<String>,
        maxPerPage: Int = 10
    ): Message? {
        event.deferReply().queue()
        val messagePages = messageLogic(event.guild!!, content, maxPerPage)
        return paginateMessage(event, messagePages)
    }

    suspend fun paginatePlaylists(
        event: SlashCommandInteractionEvent,
    ): Message? {
        event.deferReply().queue()

        val guild = event.guild!!
        val user = event.user
        val maxPerPage = 10
        val playlists = PlaylistDB.findPlaylistForUser(user.id).chunked(maxPerPage)
        require(playlists.isNotEmpty()) { "This user has no playlists to display!" }

        val pages = playlists.mapIndexed { i, chunkedPlaylists ->
            PlaylistsPage(
                guild = guild,
                _playlists = chunkedPlaylists,
                pageNumber = i
            )
        }

        return paginateImageMessage(event, pages).await()
    }

    suspend fun paginatePlaylist(
        event: SlashCommandInteractionEvent,
        playlist: PlaylistModel,
        maxTracksPerPage: Int
    ): Message? {
        event.deferReply().queue()

        val guild = event.guild!!
        val trackIndexes = mutableMapOf<PlaylistTrack, Int>()
        playlist.tracks.forEachIndexed { index, playlistTrack -> trackIndexes[playlistTrack] = index }
        val chunkedTracks = playlist.tracks.chunked(maxTracksPerPage)
        val pages = chunkedTracks.mapIndexed { i, tracks ->
            PlaylistPage(
                title = playlist.title,
                artworkUrl = playlist.artwork_url,
                description = playlist.description,
                guild = guild,
                tracks = tracks,
                trackIndexes = trackIndexes,
                pageNumber = i
            )
        }

        return paginateImageMessage(event, pages).await();
    }

    suspend fun paginateQueue(event: SlashCommandInteractionEvent, maxPerPage: Int = 10): Message? {
        val guild = event.guild!!
        val musicManager = RobertifyAudioManager[guild]
        val queueHandler = musicManager.scheduler.queueHandler

        event.deferReply().queue()
        val queuePages = messageLogic(guild, queueHandler, maxPerPage)

        return try {
            paginateQueueMessage(event, queuePages).await()
        } catch (e: ImageBuilderException) {
            val defaultMessagePages = queuePages.map { page -> DefaultMessagePage(page.getEmbed()) }
            paginateMessage(event, defaultMessagePages, true)
        }
    }

    private suspend fun paginateQueueMessage(
        event: SlashCommandInteractionEvent,
        queuePages: List<QueuePage>
    ) = paginateImageMessage(event, queuePages, true);

    private suspend fun paginateImageMessage(
        event: SlashCommandInteractionEvent,
        pages: List<AbstractImagePage>,
        isQueue: Boolean = false,
    ): Deferred<Message?> =
        coroutineScope {
            val guild = event.guild!!
            val sendFallbackEmbed: () -> Deferred<Message?> = {
                async { paginateMessage(event, pages) }
            }

            try {
                val image = pages[0].generateImage() ?: run {
                    return@coroutineScope sendFallbackEmbed()
                }

                var messageAction = event.hook
                    .sendFiles(FileUpload.fromData(image, AbstractImageBuilder.RANDOM_FILE_NAME))

                if (pages.size > 1)
                    messageAction = messageAction.addComponents(
                        paginator.getButtons(
                            user = event.user,
                            frontEnabled = false,
                            previousEnabled = false,
                            isQueue = isQueue
                        )
                    )

                val deferredMsg = async {
                    return@async messageAction.submit()
                        .thenApply { msg ->
                            if (pages.size > 1) {
                                messages[msg.idLong] = pages
                                return@thenApply msg
                            } else return@thenApply null
                        }.join()
                }

                return@coroutineScope deferredMsg
            } catch (e: Exception) {
                return@coroutineScope when (e) {
                    is SocketTimeoutException,
                    is ConnectException,
                    is ImageBuilderException -> sendFallbackEmbed()

                    else -> throw e
                }
            }
        }

    suspend fun paginateMenu(
        event: SlashCommandInteractionEvent,
        options: List<StringSelectMenuOption>,
        startingPage: Int = 0,
        numberEachEntry: Boolean = true
    ) {
        val msg = menuLogic(event, options, startingPage, numberEachEntry)
        paginateMenu(event.user, event.channel.asGuildMessageChannel(), msg, options)
    }

    fun paginateMenu(user: User, msg: Message, options: List<StringSelectMenuOption>) {
        val menuPages = menuLogic(msg.id, options)
        val firstPage = menuPages[0]
        val menu = StringSelectMenu(
            customId = "menupage:${user.id}",
            placeholder = "Select an option",
            valueRange = IntRange(1, 1),
            options = firstPage.getOptions().map { it.build() }
        )

        msg.editMessageComponents(ActionRow.of(menu))
            .queue { messages[msg.idLong] = menuPages }
    }

    fun paginateMenu(
        user: User,
        channel: GuildMessageChannel,
        msg: ReplyCallbackAction,
        options: List<StringSelectMenuOption>
    ) {
        val menuPages = menuLogic("null", options)
        val firstPage = menuPages[0]
        val menu = StringSelectMenu(
            customId = "menupage:${user.id}",
            placeholder = "Select an option",
            valueRange = IntRange(1, 1),
            options = firstPage.getOptions().map { it.build() }
        )

        msg.addActionRow(menu)
            .queue {
                it.retrieveOriginal()
                    .queue { original -> messages[original.idLong] = menuPages }
            }
    }

    fun getMessagePages(msgId: Long): List<MessagePage>? =
        messages[msgId]

    fun getQueuePages(msgId: Long): List<QueuePage>? =
        messages[msgId]?.map { it as QueuePage }

    fun getMenuPages(msgId: Long): List<MenuPage>? =
        messages[msgId]?.map { it as MenuPage }

    fun getSelectionMenu(user: User, options: List<StringSelectMenuOption>): SelectMenu =
        StringSelectionMenuBuilder(
            _name = "menuPage:${user.id}",
            placeholder = "Select an option",
            range = Pair(1, 1),
            _options = options.subList(0, options.size.coerceAtMost(InteractionLimits.SELECT_MENU_CHOICES))
        ).build()

    suspend fun getPaginatedEmbed(
        guild: Guild,
        content: List<*>,
        maxPerPage: Int = 10,
        startingPage: Int = 0,
        numberEachEntry: Boolean = true
    ): MessageEmbed {
        val embedBuilder = embedStyle().appendDescription("\t")
        var index = (startingPage * (if (maxPerPage == 0) 0 else (maxPerPage - 1))) + 1
        for (j in content.indices) {
            if (j == maxPerPage) break

            embedBuilder.appendDescription(
                "${
                    if (numberEachEntry) "**${index++}.** - " else ""
                }${content[j]}\n"
            )
        }
        return embedBuilder.build()
    }

    private suspend fun messageLogic(guild: Guild, content: List<String>, maxPerPage: Int = 10): List<DefaultMessagePage> {
        if (content.size <= maxPerPage) {
            return listOf(DefaultMessagePage(guild, content))
        } else {
            val messagePages = mutableListOf<DefaultMessagePage>()
            val pagesRequired = ceil(content.size.toDouble() / maxPerPage.toDouble()).toInt()
            var lastIndex = 0
            for (i in 0 until pagesRequired) {
                val embedBuilder = RobertifyEmbedUtils.embedMessage(guild, "\t")
                for (j in 0 until maxPerPage) {
                    if (lastIndex == content.size) break

                    embedBuilder.appendDescription("${content[lastIndex++]}\n")
                }
                messagePages.add(DefaultMessagePage(embedBuilder.build()))
            }
            return messagePages
        }
    }

    private fun messageLogic(guild: Guild, queueHandler: QueueHandler, maxPerPage: Int = 10): List<QueuePage> {
        if (queueHandler.size <= maxPerPage) {
            val items = mutableListOf<QueueItem>()
            queueHandler.contents.forEachIndexed { i, track ->
                items.add(
                    QueueItem(
                        trackIndex = i + 1,
                        trackTitle = track.info.title.coerceAtMost(30),
                        artist = track.info.author.coerceAtMost(30),
                        duration = track.length
                    )
                )
            }
            return listOf(QueuePage(guild, 1, items))
        } else {
            val messagePages = mutableListOf<QueuePage>()
            val trackList = queueHandler.contents
            val pagesRequired = ceil(queueHandler.size.toDouble() / maxPerPage.toDouble()).toInt()
            var lastIndex = 0

            for (i in 0 until pagesRequired) {
                val page = QueuePage(guild, i + 1)
                for (j in 0 until maxPerPage) {
                    if (lastIndex == queueHandler.size) break
                    val track = trackList[lastIndex]
                    page.addItem(
                        QueueItem(
                            trackIndex = lastIndex + 1,
                            trackTitle = track.info.title,
                            artist = track.info.author,
                            duration = track.length
                        )
                    )
                    lastIndex++
                }
                messagePages.add(page)
            }
            return messagePages
        }
    }

    private suspend fun menuLogic(
        event: SlashCommandInteractionEvent,
        options: List<StringSelectMenuOption>,
        startingPage: Int,
        numberEachEntry: Boolean
    ): ReplyCallbackAction =
        event.replyEmbeds(getPaginatedEmbed(event.guild!!, options, 25, startingPage, numberEachEntry))

    private fun menuLogic(msgId: String, options: List<StringSelectMenuOption>): List<MenuPage> {
        val menuPages = mutableListOf<MenuPage>()

        if (options.size <= InteractionLimits.SELECT_MENU_CHOICES) {
            val page = MenuPage()
            options.forEach { page.addOption(it) }
            menuPages.add(page)
        } else {
            val pagesRequired = ceil(options.size.toDouble() / InteractionLimits.SELECT_MENU_CHOICES.toDouble()).toInt()
            val pageControllers = 1 + ceil(((pagesRequired - 1) * 4) / 2.0).toInt()
            val actualPagesRequired =
                ceil((options.size + pageControllers).toDouble() / InteractionLimits.SELECT_MENU_CHOICES).toInt()

            var lastIndex = 0
            for (i in 0 until actualPagesRequired) {
                val tempPage = MenuPage()
                for (j in 0 until InteractionLimits.SELECT_MENU_CHOICES) {
                    if (lastIndex == options.size) break

                    if (j == 0 && i != 0) {
                        tempPage.addOption(
                            StringSelectMenuOption(
                                label = "Previous Page",
                                value = "menuPage:previousPage:${msgId}"
                            )
                        )
                        continue
                    }

                    if (j == InteractionLimits.SELECT_MENU_CHOICES - 1) {
                        tempPage.addOption(
                            StringSelectMenuOption(
                                label = "Next Page",
                                value = "menuPage:nextPage:${msgId}"
                            )
                        )
                        continue
                    }

                    tempPage.addOption(options[lastIndex++])
                }
                menuPages.add(tempPage)
            }
        }
        return menuPages
    }
}